src/lib/core/sample_sort.c

```diff
@@ -0,0 +1,513 @@
+
+	qsort_arg_st(samples, samples_num, sort->elem_size, sort->cmp,
+		     sort->cmp_arg);
+
+	/* Take splitters from samples. */
+	for (int i = 0; i < sort->threads_num - 1; i++) {
+		size_t si = oversample - 1 + i * oversample;
+		memcpy(sort->splitters + i * sort->elem_size,
+		       samples + si * sort->elem_size, sort->elem_size);
+	}
+
+	free(samples);
+}
+
+/**
+ * Check whether data part assigned to a thread is presorted.
+ */
+static int
+sample_sort_check_presorted(va_list ap)
+{
+	struct sample_sort_worker *worker = va_arg(ap, typeof(worker));
+	struct sample_sort *sort = worker->sort;
+	worker->presorted = true;
+
+	void *pos = sort->data + worker->begin * sort->elem_size;
+	void *limit = sort->data + (worker->end - 1) * sort->elem_size;
+	for (; pos < limit; pos += sort->elem_size) {
+		if (sort->cmp(pos, pos + sort->elem_size, sort->cmp_arg) > 0) {
+			worker->presorted = false;
+			break;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * Sort all the data without dividing to buckets. Used in case sorting
+ * is limited to a single thread.
+ */
+static int
+sample_sort_sort_all(va_list ap)
+{
+	struct sample_sort *sort = va_arg(ap, typeof(sort));
+
+	qsort_arg_st(sort->data, sort->data_size, sort->elem_size, sort->cmp,
+		     sort->cmp_arg);
+
+	return 0;
+}
+
+/**
+ * See usage description in header.
+ *
+ * The algorithm idea is to split the data into buckets, each element in
+ * bucket `i` is greater then each element in bucket `i - 1`, and then
+ * sort the buckets. As a result we will sort original array.
+ *
+ * So the algo outline is next:
+ *
+ * 1. Find buckets boundaries (splitters).
+ * 2. For each element in data find to what bucket is belongs to (oracles).
+ * 3. Copy elements to their buckets.
+ * 4. Sort buckets using sequentional qsort_arg and then copy buckets back
+ *    to the original array.
+ *
+ * Steps 2, 3 and 4 are run in parallel on `threads_num` threads. Step 1
+ * does not have high computation cost and is run single thread.
+ *
+ * Additionally we check if data is already sorted before applying main algo.
+ *
+ * See also:
+ * [1] https://en.wikipedia.org/wiki/Samplesort
+ * [2] Super Scalar Sample Sort, Peter Sanders and Sebastian Winkel
+ *
+ * Although this implementation does not use superscalar approach to map
+ * elements to buckets as in [2] and use usual binary search.
+ */
+void
+sample_cosort_mt(void *data, size_t data_size, size_t elem_size,
+		 compare_arg_f cmp, void *cmp_arg, int threads_num)
+{
+	struct sample_sort sort;
+	double b, e;
+
+	say_verbose("start sort, data size: %zd, elem size: %zd, threads: %d\n",
+		    data_size, elem_size, threads_num);
+	if (data_size < NOSPAWN_SIZE_THESHOLD) {
+		say_verbose("data size is less then threshold %d,"
+			    " sort in caller thread", NOSPAWN_SIZE_THESHOLD);
+		qsort_arg_st(data, data_size, elem_size, cmp, cmp_arg);
+		return;
+	}
+
+	/* Upper limit is because oracles are chars. */
+	assert(threads_num > 0 && threads_num <= SAMPLE_SORT_THREADS_MAX);
```

Comment:
Discussed f2f and decided to keep this assertion. Rewored the comment a bit:
```c
/*
 * Upper limit because element to bucket map has unsigned char
 * storage per element.
 */
```